### PR TITLE
Polynomial Regression Predictor

### PR DESCRIPTION
**Pull request by davidrichey**
* Added Polynomial Regression Prediction

Note: dependency [Matrix](https://hex.pm/packages/matrix) added for Matrix math

### CODE REVIEW

File: test/learn_kit/regression/polynomial_test.exs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  test "create new polynomial predictor with data" do
+    assert Polynomial.new([1, 2], [3, 4]) == %Polynomial{
+             coefficients: [],
+             degree: 2,
+             factors: [1, 2],
+             results: [3, 4]
+           }
+  end
+
+  def factors, do: [-3, -2, -1, -0.2, 1, 3]
```
Comment (kortirso) [inline test/learn_kit/regression/polynomial_test.exs]: maybe use setup_all?

File: lib/learn_kit/regression/polynomial.ex [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+  ## Examples
+
+      iex> predictor |> LearnKit.Regression.Polynomial.predict([5,6])
+      {:ok, [20.999999999999723, 27.999999999999574]}
+
+  """
+  def predict(polynomial = %Polynomial{coefficients: _, degree: _}, samples)
+      when is_list(samples) do
+    {:ok,
+     Enum.map(samples, fn sample ->
```
Comment (kortirso) [inline lib/learn_kit/regression/polynomial.ex]: can you add here private function do_predict for map?

File: lib/learn_kit/regression/polynomial.ex [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+      Enum.map(power_ofs, fn factor ->
+        sum_x_with_k(factors, factor, 0.0)
+      end)
+  end
+
+  defp matrix_line(line, factors, degree) do
+    line_factor = line - 1
+    power_ofs = Enum.to_list(line_factor..(degree + line_factor))
+
+    Enum.map(power_ofs, fn factor ->
```
Comment (kortirso) [inline lib/learn_kit/regression/polynomial.ex]: code duplication

File: lib/learn_kit/regression/polynomial.ex [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+  """
+  def predict(%Polynomial{coefficients: coefficients, degree: degree}, sample) do
+    ordered_coefficients = coefficients |> Enum.reverse()
+    {:ok, substitute_coefficients(ordered_coefficients, sample, degree, 0.0)}
+  end
+
+  defp matrix_line(1, factors, degree) do
+    power_ofs = Enum.to_list(1..degree)
+
+    [Enum.count(factors)] ++
```
Comment (kortirso) [inline lib/learn_kit/regression/polynomial.ex]: Is it possible to use here prepending an element to a list, because it's faster then ++?

**Comment by kortirso** [comment]
@davidrichey Thank you for PR, but 1 have a few questions
- what is better practice: left private functions in core module or hide them in additional?
- you declared @types, but there are no specs
- is it possible to add calculation of the coefficient of determination R^2 of the prediction?
- and there are a few comments in the code

**Comment by davidrichey** [comment]
@kortirso thank you for the detailed comments and questions.
I have update this PR with a couple commits:
* First [Formatter Updates of Files I intend to change](https://github.com/kortirso/elixir_learn_kit/pull/3/commits/7e2ae5e6e72370718e3a276cc688d3a905ff54a8)
* Next the [Updates from your comments](https://github.com/kortirso/elixir_learn_kit/pull/3/commits/9747ebe0a17e1c7404dc9d608186be58d4738cfd) - I have pulled some on the calculations out of the `Linear.Calculations` as it will be applied to the Polynomial regression error calculation
* And last, [added polynomial regression scoring](https://github.com/kortirso/elixir_learn_kit/pull/3/commits/3e408eca8bc5c3932b94c12668430dfaee69b1e6)

**Comment by kortirso** [comment]
@davidrichey Merged, thank you for your work

**Pull request merged**